### PR TITLE
[SPARK-31665][SQL][TESTS] Check parquet dictionary encoding of random dates/timestamps

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Modified `RandomDataGenerator.forType` for DateType and TimestampType to generate special date//timestamp values with 0.5 probability. This will trigger dictionary encoding in Parquet datasource test  HadoopFsRelationTest "test all data types". Currently, dictionary encoding is tested only for numeric types like ShortType.

### Why are the changes needed?
To extend test coverage. Currently, probability of testing of dictionary encoding in the test HadoopFsRelationTest "test all data types" for DateType and TimestampType is close to zero because dates/timestamps are uniformly distributed in wide range, and the chance of generating the same values is pretty low. In this way, parquet datasource cannot apply dictionary encoding for such column types.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
By running `ParquetHadoopFsRelationSuite` and `JsonHadoopFsRelationSuite`.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/HadoopFsRelationTest.scala [hunk 1]
```diff
@@ -159,7 +159,7 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils with Tes
               .add("index", IntegerType, nullable = false)
               .add("col", dataType, nullable = true)
             val rdd =
-              spark.sparkContext.parallelize((1 to 10).map(i => Row(i, dataGenerator())))
+              spark.sparkContext.parallelize((1 to 20).map(i => Row(i, dataGenerator())))
```
Comment (MaxGekk) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/HadoopFsRelationTest.scala:162]: I bumped the number to have the same amount of random values as before the changes.

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon @srowen @dongjoon-hyun May I ask you to review this PR.

**Comment by MaxGekk** [comment]
This is some kind of companion checks for https://github.com/apache/spark/pull/28479 when date rebasing is off.

**Comment by srowen** [comment]
Seems OK if it's just improving test coverage

**Comment by MaxGekk** [comment]
The roundtrip test RowEncoderSuite.encode/decode fails https://github.com/apache/spark/pull/28481#issuecomment-626034381 on 1000-02-29 because the date doesn't exist in Proleptic Gregorian calendar, and Spark shifts it to the next valid date 1000-03-01. The test didn't fail before so often because the probability of generating this date was pretty low.

**Comment by MaxGekk** [comment]
Here is JIRA ticket for [the test failure](https://github.com/apache/spark/pull/28481#issuecomment-626034381) https://issues.apache.org/jira/browse/SPARK-31669

**Comment by MaxGekk** [comment]
I will exclude the date 1000-02-29 because:
1. The test RowEncoderSuite.encode/decode can fail before this PR
2. The main goal of this PR is to test dictionary encoding that is triggered when there are duplicates in DATE/TIMESTAMP columns

**Comment by MaxGekk** [comment]
The PR https://github.com/apache/spark/pull/28486 fixes the build failure https://github.com/apache/spark/pull/28481#issuecomment-626034381

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala [hunk 2]
```diff
@@ -162,52 +162,66 @@ object RandomDataGenerator {
...
-            // a local date in Proleptic Gregorian calendar to satisfy this requirement.
-            // Some years are leap years in Julian calendar but not in Proleptic Gregorian calendar.
-            // As the consequence of that, 29 February of such years might not exist in Proleptic
-            // Gregorian calendar. When this happens, we shift the date by one day.
-            Try { date.toLocalDate; date }.getOrElse(new Date(date.getTime + MILLIS_PER_DAY))
+        def uniformDateRand(rand: Random): java.sql.Date = {
+          var milliseconds = rand.nextLong() % 253402329599999L
+          // -62135740800000L is the number of milliseconds before January 1, 1970, 00:00:00 GMT
+          // for "0001-01-01 00:00:00.000000". We need to find a
+          // number that is greater or equals to this number as a valid timestamp value.
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala:169]: I'm a bit unsure about this. Spark does support negative years for now, and we should keep testing it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala [hunk 3]
```diff
@@ -162,52 +162,66 @@ object RandomDataGenerator {
...
+          while (milliseconds < -62135740800000L) {
+            // 253402329599999L is the number of milliseconds since
+            // January 1, 1970, 00:00:00 GMT for "9999-12-31 23:59:59.999999".
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala:172]: ditto.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala:169]: Officially, not. See descriptions of TimestampType and DateType.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala:172]: It is not related to this PR. If you believe we should change the lower and upper bounds, we should do that separately.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala:169]: oh sorry, the upper/lower bound is already here. Then I'm fine

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Pull request closed without merging**